src/main/java/org/jenkinsci/plugins/mesos/MesosApi.java

```diff
@@ -49,18 +65,24 @@ public MesosApi(String masterUrl, String user, String frameworkName)
             .withValue("master-url", ConfigValueFactory.fromAnyRef(masterUrl));
     this.clientSettings = MesosClientSettings.fromConfig(conf);
     system = ActorSystem.create("mesos-scheduler");
+    context = system.dispatcher();
     materializer = ActorMaterializer.create(system);
-
     client = connectClient().get();
+    stateMap = new ConcurrentHashMap<>();
+    adapter = new SchedulerAdapter(Scheduler.fromClient(client), materializer, context);
+    adapterSourceSink = adapter.asSourceAndSink(SpecsSnapshot.empty());
+    adapterSourceSink.t2().runForeach(this::updateState, materializer);
+    adapterQueues = adapter.asAkkaQueues(SpecsSnapshot.empty(), OverflowStrategy.backpressure());
   }
 
   /**
-   * Start a Jekins slave/agent on Mesos.
-   *
-   * @return A future reference to the {@link MesosSlave}.
+   * Enqueue spec for a jenkins agent that will eventually come online.
    */
-  public CompletableFuture<MesosSlave> startAgent() {
-    throw new NotImplementedException();
+  public void enqueueAgent() {
+    PodSpec spec = buildMesosAgentTask(0.1, 32);
+    SpecUpdated update = new PodSpecUpdated(spec.id(), Option.apply(spec));
+    //async add agent to queue
+    adapterQueues.t2().offer(update);
```

Comment:
Don't we want to return a future or at least block here in order to know when we're allowed to call the scheduler again? If the `offer` method is called second time before the future is completed, it will result in an error.

Reply:
I thought `offer` is blocking similar to a blocking queue.

Reply:
No, it returns `Future[QueueOfferResult]` (so no blocking ;) ) see, how we use it in the [InstanceTrackerDelegate](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala#L119)